c10/core/impl/SizesAndStrides.h

```diff
@@ -73,46 +194,120 @@ class C10_API SizesAndStrides {
 
   // Size accessors.
   int64_t size_at(size_t idx) const {
-    return sizes_.at(idx);
+    assert(idx < size());
+    return sizes_data()[idx];
   }
 
   int64_t& size_at(size_t idx) {
-    return sizes_.at(idx);
+    assert(idx < size());
+    return sizes_data()[idx];
   }
 
   int64_t size_at_unchecked(size_t idx) const {
-    return sizes_[idx];
+    return sizes_data()[idx];
   }
 
   int64_t& size_at_unchecked(size_t idx) {
-    return sizes_[idx];
+    return sizes_data()[idx];
   }
 
   // Size accessors.
   int64_t stride_at(size_t idx) const {
-    return strides_.at(idx);
+    assert(idx < size());
+    return strides_data()[idx];
   }
 
   int64_t& stride_at(size_t idx) {
-    return strides_.at(idx);
+    assert(idx < size());
+    return strides_data()[idx];
   }
 
   int64_t stride_at_unchecked(size_t idx) const {
-    return strides_[idx];
+    return strides_data()[idx];
   }
 
   int64_t& stride_at_unchecked(size_t idx) {
-    return strides_[idx];
+    return strides_data()[idx];
   }
 
-  void resize(size_t sz) {
-    sizes_.resize(sz);
-    strides_.resize(sz);
+  void resize(const size_t newSize) {
+    const auto oldSize = size();
+    if (newSize == oldSize) {
+      return;
+    }
+    if (C10_LIKELY(newSize <= MAX_INLINE_SIZE)) {
+      if (C10_LIKELY(isInline())) {
+        if (oldSize < newSize) {
+          const auto bytesToZero = (newSize - oldSize) * sizeof(inlineStorage_[0]);
+          memset(&inlineStorage_[oldSize], 0, bytesToZero);
+          memset(&inlineStorage_[MAX_INLINE_SIZE + oldSize], 0, bytesToZero);
+        }
```

Comment:
In debug mode we should zero/sentinel out the slop, make it easier to figure out if something has gone wrong.

Reply:
I would advocate for doing it in release mode too. Stuff being zero-initialized in debug mode but not in release mode is a long standing issue in some C++ compilers that makes debugging hard since stuff breaks when you run it in release mode and then suddenly it works when you rerun in debug mode.

Reply:
I'd rather not add unnecessary memory references in release builds. AddressSanitizer will help with use of uninitialized memory already, and we can add test coverage to make sure things are working, so I've opted for not adding this debugging feature.